File: mongodb-learning/src/main/java/com/harshit/mongodb/basic/Example2CRUD.java
package com.harshit.mongodb.basic;

import com.harshit.mongodb.config.DatabaseConfig;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.Filters;
import com.mongodb.client.model.Updates;
import org.bson.Document;
import org.bson.types.ObjectId;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;

/**
 * Example 2: CRUD Operations (Create, Read, Update, Delete)
 * 
 * SIMPLE EXPLANATION:
 * CRUD stands for the four basic operations you can do with data:
 * - CREATE: Add new data
 * - READ: Get/view data
 * - UPDATE: Change existing data
 * - DELETE: Remove data
 * 
 * TECHNICAL EXPLANATION:
 * MongoDB provides methods for each CRUD operation:
 * - Create: insertOne(), insertMany()
 * - Read: findOne(), find()
 * - Update: updateOne(), updateMany(), replaceOne()
 * - Delete: deleteOne(), deleteMany()
 * 
 * INTERVIEW POINT:
 * - insertOne returns InsertOneResult with insertedId
 * - find() returns a FindIterable, use into() to convert to list
 * - updateOne updates first match, updateMany updates all matches
 * - Always use $set operator in updates to avoid replacing entire document
 */
public class Example2CRUD {
    
    private static final Logger logger = LoggerFactory.getLogger(Example2CRUD.class);
    
    public static void main(String[] args) {
        MongoClient client = null;
        
        try {
            logger.info("🔄 Connecting to MongoDB...\n");
            MongoDatabase db = DatabaseConfig.connectToDatabase();
            client = DatabaseConfig.getClient();
            
            MongoCollection<Document> usersCollection = db.getCollection("users");
            
            // ============================================
            // CREATE OPERATIONS
            // ============================================
            logger.info("📝 === CREATE OPERATIONS ===\n");
            
            // Insert a single document
            logger.info("1. Inserting a single user...");
            Document newUser = new Document()
                    .append("name", "John Doe")
                    .append("email", "john.doe@example.com")
                    .append("age", 30)
                    .append("city", "New York")
                    .append("interests", Arrays.asList("coding", "reading", "traveling"))
                    .append("createdAt", new Date());
            
            usersCollection.insertOne(newUser);
            ObjectId johnId = newUser.getObjectId("_id");
            logger.info("   ✅ User inserted with ID: {}\n", johnId);
            
            // Insert multiple documents
            logger.info("2. Inserting multiple users...");
            List<Document> multipleUsers = Arrays.asList(
                    new Document()
                            .append("name", "Jane Smith")
                            .append("email", "jane.smith@example.com")
                            .append("age", 25)
                            .append("city", "Los Angeles")
                            .append("interests", Arrays.asList("music", "dancing"))
                            .append("createdAt", new Date()),
                    new Document()
                            .append("name", "Bob Johnson")
                            .append("email", "bob.johnson@example.com")
                            .append("age", 35)
                            .append("city", "Chicago")
                            .append("interests", Arrays.asList("sports", "cooking"))
                            .append("createdAt", new Date()),
                    new Document()
                            .append("name", "Alice Williams")
                            .append("email", "alice.williams@example.com")
                            .append("age", 28)
                            .append("city", "New York")
                            .append("interests", Arrays.asList("art", "photography", "traveling"))
                            .append("createdAt", new Date())
            );
            
            usersCollection.insertMany(multipleUsers);
            logger.info("   ✅ {} users inserted\n", multipleUsers.size());
            
            // ============================================
            // READ OPERATIONS
            // ============================================
            logger.info("📖 === READ OPERATIONS ===\n");
            
            // Find one document
            logger.info("1. Finding one user by email...");
            Document foundUser = usersCollection.find(Filters.eq("email", "john.doe@example.com")).first();
            if (foundUser != null) {
                logger.info("   Found user: {}", foundUser.toJson());
            }
            logger.info("");
            
            // Find all documents
            logger.info("2. Finding all users...");
            List<Document> allUsers = usersCollection.find().into(new ArrayList<>());
            logger.info("   Total users: {}", allUsers.size());
            for (int i = 0; i < allUsers.size(); i++) {
                Document user = allUsers.get(i);
                logger.info("   {}. {} ({})", i + 1, user.getString("name"), user.getString("email"));
            }
            logger.info("");
            
            // Find with filter
            logger.info("3. Finding users from New York...");
            List<Document> nyUsers = usersCollection.find(Filters.eq("city", "New York")).into(new ArrayList<>());
            logger.info("   Found {} users from New York:", nyUsers.size());
            nyUsers.forEach(user -> {
                logger.info("   - {}", user.getString("name"));
            });
            logger.info("");
            
            // Find with condition (age > 30)
            logger.info("4. Finding users older than 30...");
            List<Document> olderUsers = usersCollection.find(Filters.gt("age", 30)).into(new ArrayList<>());
            logger.info("   Found {} users older than 30:", olderUsers.size());
            olderUsers.forEach(user -> {
                logger.info("   - {} (age: {})", user.getString("name"), user.getInteger("age"));
            });
            logger.info("");
            
            // ============================================
            // UPDATE OPERATIONS
            // ============================================
            logger.info("✏️  === UPDATE OPERATIONS ===\n");
            
            // Update one document
            logger.info("1. Updating John's age to 31...");
            usersCollection.updateOne(
                    Filters.eq("email", "john.doe@example.com"),
                    Updates.combine(
                            Updates.set("age", 31),
                            Updates.set("updatedAt", new Date())
                    )
            );
            logger.info("   ✅ Document updated\n");
            
            // Verify update
            Document updatedUser = usersCollection.find(Filters.eq("email", "john.doe@example.com")).first();
            if (updatedUser != null) {
                logger.info("   Updated age: {}\n", updatedUser.getInteger("age"));
            }
            
            // Update many documents (add field to all users)
            logger.info("2. Adding \"status\" field to all users...");
            long modifiedCount = usersCollection.updateMany(
                    new Document(),
                    Updates.set("status", "active")
            ).getModifiedCount();
            logger.info("   ✅ {} documents updated\n", modifiedCount);
            
            // Increment a field
            logger.info("3. Incrementing Bob's age by 1...");
            usersCollection.updateOne(
                    Filters.eq("email", "bob.johnson@example.com"),
                    Updates.inc("age", 1)
            );
            Document bob = usersCollection.find(Filters.eq("email", "bob.johnson@example.com")).first();
            if (bob != null) {
                logger.info("   Bob's new age: {}\n", bob.getInteger("age"));
            }
            
            // ============================================
            // DELETE OPERATIONS
            // ============================================
//            logger.info("🗑️  === DELETE OPERATIONS ===\n");
//
//            // Delete one document
//            logger.info("1. Deleting Alice...");
//            long deletedCount = usersCollection.deleteOne(Filters.eq("email", "alice.williams@example.com")).getDeletedCount();
//            logger.info("   ✅ {} document deleted\n", deletedCount);
//
//            // Verify deletion
//            List<Document> remainingUsers = usersCollection.find().into(new ArrayList<>());
//            logger.info("   Remaining users: {}\n", remainingUsers.size());
//
//            // Delete many documents (cleanup - optional)
//            logger.info("2. Cleaning up test data...");
//            long deletedManyCount = usersCollection.deleteMany(new Document()).getDeletedCount();
//            logger.info("   ✅ {} documents deleted\n", deletedManyCount);
            
            logger.info("✅ CRUD operations completed successfully!");
            
        } catch (Exception error) {
            logger.error("❌ Error: {}", error.getMessage(), error);
        } finally {
            if (client != null) {
                DatabaseConfig.closeConnection();
            }
            // Give MongoDB driver threads more time to fully terminate
            try {
                Thread.sleep(500);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            // Force JVM exit to prevent thread cleanup warnings with Maven exec plugin
            System.exit(0);
        }
    }
}

File: mongodb-learning/basic/Example2-CRUD.js
/**
 * Example 2: CRUD Operations (Create, Read, Update, Delete)
 * 
 * SIMPLE EXPLANATION:
 * CRUD stands for the four basic operations you can do with data:
 * - CREATE: Add new data
 * - READ: Get/view data
 * - UPDATE: Change existing data
 * - DELETE: Remove data
 * 
 * TECHNICAL EXPLANATION:
 * MongoDB provides methods for each CRUD operation:
 * - Create: insertOne(), insertMany()
 * - Read: findOne(), find()
 * - Update: updateOne(), updateMany(), replaceOne()
 * - Delete: deleteOne(), deleteMany()
 * 
 * INTERVIEW POINT:
 * - insertOne returns InsertOneResult with insertedId
 * - find() returns a cursor, use toArray() or forEach() to iterate
 * - updateOne updates first match, updateMany updates all matches
 * - Always use $set operator in updates to avoid replacing entire document
 */

const { connectToDatabase, closeConnection } = require('../config/database');

async function main() {
    let client;
    
    try {
        console.log('🔄 Connecting to MongoDB...\n');
        const { db, client: dbClient } = await connectToDatabase();
        client = dbClient;
        
        const usersCollection = db.collection('users');
        
        // ============================================
        // CREATE OPERATIONS
        // ============================================
        console.log('📝 === CREATE OPERATIONS ===\n');
        
        // Insert a single document
        console.log('1. Inserting a single user...');
        const newUser = {
            name: 'John Doe',
            email: 'john.doe@example.com',
            age: 30,
            city: 'New York',
            interests: ['coding', 'reading', 'traveling'],
            createdAt: new Date()
        };
        
        const insertResult = await usersCollection.insertOne(newUser);
        console.log(`   ✅ User inserted with ID: ${insertResult.insertedId}\n`);
        
        // Insert multiple documents
        console.log('2. Inserting multiple users...');
        const multipleUsers = [
            {
                name: 'Jane Smith',
                email: 'jane.smith@example.com',
                age: 25,
                city: 'Los Angeles',
                interests: ['music', 'dancing'],
                createdAt: new Date()
            },
            {
                name: 'Bob Johnson',
                email: 'bob.johnson@example.com',
                age: 35,
                city: 'Chicago',
                interests: ['sports', 'cooking'],
                createdAt: new Date()
            },
            {
                name: 'Alice Williams',
                email: 'alice.williams@example.com',
                age: 28,
                city: 'New York',
                interests: ['art', 'photography', 'traveling'],
                createdAt: new Date()
            }
        ];
        
        const insertManyResult = await usersCollection.insertMany(multipleUsers);
        console.log(`   ✅ ${insertManyResult.insertedCount} users inserted\n`);
        
        // ============================================
        // READ OPERATIONS
        // ============================================
        console.log('📖 === READ OPERATIONS ===\n');
        
        // Find one document
        console.log('1. Finding one user by email...');
        const foundUser = await usersCollection.findOne({ email: 'john.doe@example.com' });
        console.log('   Found user:', JSON.stringify(foundUser, null, 2));
        console.log('');
        
        // Find all documents
        console.log('2. Finding all users...');
        const allUsers = await usersCollection.find({}).toArray();
        console.log(`   Total users: ${allUsers.length}`);
        allUsers.forEach((user, index) => {
            console.log(`   ${index + 1}. ${user.name} (${user.email})`);
        });
        console.log('');
        
        // Find with filter
        console.log('3. Finding users from New York...');
        const nyUsers = await usersCollection.find({ city: 'New York' }).toArray();
        console.log(`   Found ${nyUsers.length} users from New York:`);
        nyUsers.forEach(user => {
            console.log(`   - ${user.name}`);
        });
        console.log('');
        
        // Find with condition (age > 30)
        console.log('4. Finding users older than 30...');
        const olderUsers = await usersCollection.find({ age: { $gt: 30 } }).toArray();
        console.log(`   Found ${olderUsers.length} users older than 30:`);
        olderUsers.forEach(user => {
            console.log(`   - ${user.name} (age: ${user.age})`);
        });
        console.log('');
        
        // ============================================
        // UPDATE OPERATIONS
        // ============================================
        console.log('✏️  === UPDATE OPERATIONS ===\n');
        
        // Update one document
        console.log('1. Updating John\'s age to 31...');
        const updateResult = await usersCollection.updateOne(
            { email: 'john.doe@example.com' },
            { $set: { age: 31, updatedAt: new Date() } }
        );
        console.log(`   ✅ ${updateResult.modifiedCount} document updated\n`);
        
        // Verify update
        const updatedUser = await usersCollection.findOne({ email: 'john.doe@example.com' });
        console.log(`   Updated age: ${updatedUser.age}\n`);
        
        // Update many documents (add field to all users)
        console.log('2. Adding "status" field to all users...');
        const updateManyResult = await usersCollection.updateMany(
            {},
            { $set: { status: 'active' } }
        );
        console.log(`   ✅ ${updateManyResult.modifiedCount} documents updated\n`);
        
        // Increment a field
        console.log('3. Incrementing Bob\'s age by 1...');
        await usersCollection.updateOne(
            { email: 'bob.johnson@example.com' },
            { $inc: { age: 1 } }
        );
        const bob = await usersCollection.findOne({ email: 'bob.johnson@example.com' });
        console.log(`   Bob's new age: ${bob.age}\n`);
        
        // ============================================
        // DELETE OPERATIONS
        // ============================================
        console.log('🗑️  === DELETE OPERATIONS ===\n');
        
        // Delete one document
        console.log('1. Deleting Alice...');
        const deleteResult = await usersCollection.deleteOne({ email: 'alice.williams@example.com' });
        console.log(`   ✅ ${deleteResult.deletedCount} document deleted\n`);
        
        // Verify deletion
        const remainingUsers = await usersCollection.find({}).toArray();
        console.log(`   Remaining users: ${remainingUsers.length}\n`);
        
        // Delete many documents (cleanup - optional)
        console.log('2. Cleaning up test data...');
        const deleteManyResult = await usersCollection.deleteMany({});
        console.log(`   ✅ ${deleteManyResult.deletedCount} documents deleted\n`);
        
        console.log('✅ CRUD operations completed successfully!');
        
    } catch (error) {
        console.error('❌ Error:', error.message);
        console.error(error);
    } finally {
        if (client) {
            await closeConnection(client);
        }
    }
}

// Run the example
main();

